blockchain/utxocache.go

```diff
@@ -0,0 +1,790 @@
+	// the additional overhead of using a read-write mutex.
+	cacheLock sync.Mutex
+
+	// entries holds the cached utxo entries.
+	entries map[wire.OutPoint]*UtxoEntry
+
+	// lastFlushHash is the block hash of the last flush.  It is used to compare
+	// the state of the cache to the utxo set state in the database so that the
+	// utxo set can properly be initialized in the case that the latest utxo data
+	// had not been flushed to the database yet.
+	lastFlushHash chainhash.Hash
+
+	// lastFlushTime is the last time that the cache was flushed to the database.
+	// It is used to determine when to periodically flush the cache to the
+	// database during initial block download even if the cache isn't full to
+	// minimize the amount of progress lost if an unclean shutdown occurs.
+	lastFlushTime time.Time
+
+	// lastEvictionHeight is the block height of the last eviction.  When the
+	// cache reaches the maximum allowed size, entries are evicted based on the
+	// height of the block that they are contained in, and last eviction height is
+	// updated to the current height.
+	lastEvictionHeight uint32
+
+	// totalEntrySize is the total size of all utxo entries in the cache, in
+	// bytes.  It is updated whenever an entry is added or removed from the cache.
+	totalEntrySize uint64
+
+	// The following fields track the total number of cache hits and misses and
+	// are used to measure the overall cache hit ratio.
+	hits   uint64
+	misses uint64
+}
+
+// UtxoCacheConfig is a descriptor which specifies the utxo cache instance
+// configuration.
+type UtxoCacheConfig struct {
+	// DB defines the database which houses the utxo set.
+	//
+	// This field is required.
+	DB database.DB
+
+	// MaxSize defines the maximum allowed size of the utxo cache, in bytes.
+	//
+	// This field is required.
+	MaxSize uint64
+}
+
+// NewUtxoCache returns a UtxoCache instance using the provided configuration
+// details.
+func NewUtxoCache(config *UtxoCacheConfig) *UtxoCache {
+	// Approximate the maximum number of entries allowed in the cache in order
+	// to set the initial capacity of the entries map.
+	avgEntrySize := outpointSize + pointerSize + baseEntrySize + p2pkhScriptLen
+	maxEntries := math.Ceil(float64(config.MaxSize) / float64(avgEntrySize))
+
+	return &UtxoCache{
+		db:            config.DB,
+		maxSize:       config.MaxSize,
+		entries:       make(map[wire.OutPoint]*UtxoEntry, uint64(maxEntries)),
+		lastFlushTime: time.Now(),
+	}
+}
+
+// totalSize returns the total size of the cache on a 64-bit platform, in bytes.
+// Note that this only takes the entries map into account, which represents the
+// vast majoirty of the memory that the cache uses, and does not include the
+// memory usage of other fields in the utxo cache struct.
+//
+// This function MUST be called with the cache lock held.
+func (c *UtxoCache) totalSize() uint64 {
+	numEntries := uint64(len(c.entries))
+	return outpointSize*numEntries + pointerSize*numEntries + c.totalEntrySize
```

Comment:
This probably should try to take into account the map overhead since it is fairly significant.

For example, here is the output of some test code I wrote using `runtime/metrics` to calculate the per item usage of a map with outpoints as keys and values of just a pointer to a `uint32`.  Given that, you would expect `56 + 8 + 4 = 68` bytes per entry with the rest being the map overhead (internal buckets and such).  So, as can be seen, that results in an overhead anywhere between 37 to 196 bytes depending on the number of entries.

```
map with 56-byte outpoint keys  -- entries: 100000, size:  10511176 B, per item: 105.11 B
map with 56-byte outpoint keys  -- entries: 200000, size:  21103912 B, per item: 105.52 B
map with 56-byte outpoint keys  -- entries: 300000, size:  37966888 B, per item: 126.56 B
map with 56-byte outpoint keys  -- entries: 400000, size:  42206384 B, per item: 105.52 B
map with 56-byte outpoint keys  -- entries: 500000, size:  75529248 B, per item: 151.06 B
map with 56-byte outpoint keys  -- entries: 600000, size:  75929304 B, per item: 126.55 B
map with 56-byte outpoint keys  -- entries: 700000, size:  78671896 B, per item: 112.39 B
map with 56-byte outpoint keys  -- entries: 800000, size:  84432744 B, per item: 105.54 B
map with 56-byte outpoint keys  -- entries: 900000, size: 228515544 B, per item: 253.91 B
map with 56-byte outpoint keys  -- entries: 1000000, size: 151060952 B, per item: 151.06 B
map with 56-byte outpoint keys  -- entries: 2000000, size: 302123536 B, per item: 151.06 B
map with 56-byte outpoint keys  -- entries: 3000000, size: 325125824 B, per item: 108.38 B
map with 56-byte outpoint keys  -- entries: 4000000, size: 604249056 B, per item: 151.06 B
map with 56-byte outpoint keys  -- entries: 5000000, size: 608249264 B, per item: 121.65 B
map with 56-byte outpoint keys  -- entries: 6000000, size: 650438448 B, per item: 108.41 B
map with 56-byte outpoint keys  -- entries: 7000000, size: 1852802320 B, per item: 264.69 B
map with 56-byte outpoint keys  -- entries: 8000000, size: 1208500256 B, per item: 151.06 B
map with 56-byte outpoint keys  -- entries: 9000000, size: 1212500400 B, per item: 134.72 B
```

Discarding two big outliers, it looks like the average overhead over the entire range is about 57 bytes.

In other words, I expect all of these calculations (e.g. including the `avgEntrySize` in `NewUtxoCache`) should probably add an additional term for an overhead of ` numEntries*57` bytes to account for it and that should get the actual overall memory usage closer to the configured value.

The rest of the extra memory usage as compared to what might be expected, as previously mentioned, is very likely due to the caching in the database layer as well as the GC over provisioning when a barrage of allocations all arrive at once.

Reply:
Thanks for running these numbers, I think this is a great suggestion.  I updated to use the 57 bytes of map overhead per entry.

With this change, I also updated the default `utxocachemaxsize` from 100 MiB to 150 MiB, which keeps the overall memory usage and hit rate approximately the same.  Adding the extra 57 bytes per entry to the average of 145 bytes per entry I was using before is 57/145 or about a ~39% increase.  To keep things about the same, we'd need to increase the default size by about that much, which would bring the default up to 139 MiB, so I just rounded up to 150 MiB to use a nicer number.  I also confirmed that with these changes the total memory usage after a full sync still peaks at about 1.25GB.